02_knight_and_princess/roman_kuprii/main.cpp

```diff
@@ -0,0 +1,131 @@
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using std::endl;
+using std::to_string;
+using std::vector;
+
+using EdgeId = int;
+
+struct Edge {
+  EdgeId id;
+  std::array<int, 2> nodes;
+
+  Edge(int start, int end, int _id) : id(_id) {
+    nodes[0] = start;
+    nodes[1] = end;
+  }
+
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"id\": ";
+    res += to_string(id);
+    res += ", \"vertex_ids\": [";
+    res += to_string(nodes[0]);
+    res += ", ";
+    res += to_string(nodes[1]);
+    res += "] }";
+    return res;
+  }
+};
+
+// TODO check uniqueness of ID field
+struct Vertex {
+  int id = -1;
+  std::vector<int> edges_ids;
+
+  explicit Vertex(int _id) : id(_id) {}
+
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"id\": ";
+    res += to_string(id) + ", \"edge_ids\": [";
+    for (int n : edges_ids) {
+      res += to_string(n);
+      res += ", ";
+    }
+    res.pop_back();
+    res.pop_back();
+    res += "] }";
+    return res;
+  }
+};
+
+class Graph {
+ public:
+  Graph(const vector<Edge>& init_edges, const vector<Vertex>& init_vertices)
+      : vertices_(init_vertices), edges_(init_edges), size_(vertices_.size()) {}
```

Comment:
Усложняешь себе жизнь с таким конструктором. Лучше бы `Graph` сам отвечал за добавление вершин и граней, можно было бы избежать сложной валидирующей логики, но хозяин - барин)

В таком случае, тебе нужно удостовериться что входные параметры имеют смысл. Соотвественно грани должны соединять вершины, которые сюда передаются. Иначе они могут соединять несуществующие вершины. И второе, все вершины должны быть соединены, нужно проверить что тут нет вершин без граней.
```suggestion
  Graph(const vector<Edge>& init_edges, const vector<Vertex>& init_vertices)
      : vertices_(init_vertices), edges_(init_edges), size_(vertices_.size()) {
      // validation  
    }
```